himbaechel/uarch/gatemate/route_mult.cc

```diff
@@ -0,0 +1,826 @@
+    auto x2y2 = ctx->idf("X%dY%d", loc.x + 1, loc.y + 1);
+
+    auto cpe_combout2 = ctx->getBelPinWire(upper->bel, id_OUT);
+    auto cpe_out2_int = ctx->getWireByName(IdStringList::concat(x1y1, ctx->idf("CPE.OUT2_int")));
+    auto in_mux = ctx->getWireByName(IdStringList::concat(x2y2, ctx->idf("IM.P08.D1")));
+
+    if (bind_route_start) {
+        ctx->bindWire(cpe_combout2, net, STRENGTH_LOCKED);
+        find_and_bind_downhill_pip(ctx, cpe_combout2, cpe_out2_int, net);
+        if (is_fourgroup_a) {
+            auto sb_big_d0 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_BIG.P08.D0")));
+            find_and_bind_downhill_pip(ctx, cpe_out2_int, sb_big_d0, net);
+        } else {
+            auto sb_sml_d0 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_SML.P08.D0")));
+            find_and_bind_downhill_pip(ctx, cpe_out2_int, sb_sml_d0, net);
+        }
+    }
+
+    if (is_fourgroup_a) {
+        auto sb_big_d0 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_BIG.P08.D0")));
+        auto sb_big_y1 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_BIG.P08.Y1")));
+        auto sb_sml_y1_int = ctx->getWireByName(IdStringList::concat(x2y0, ctx->idf("SB_SML.P08.Y1_int")));
+        auto sb_sml_ydiag_int = ctx->getWireByName(IdStringList::concat(x2y0, ctx->idf("SB_SML.P08.YDIAG_int")));
+        auto sb_sml_y2_int = ctx->getWireByName(IdStringList::concat(x2y0, ctx->idf("SB_SML.P08.Y2_int")));
+
+        find_and_bind_downhill_pip(ctx, sb_big_d0, sb_big_y1, net);
+        find_and_bind_downhill_pip(ctx, sb_big_y1, sb_sml_y1_int, net);
+        find_and_bind_downhill_pip(ctx, sb_sml_y1_int, sb_sml_ydiag_int, net);
+        find_and_bind_downhill_pip(ctx, sb_sml_ydiag_int, sb_sml_y2_int, net);
+        find_and_bind_downhill_pip(ctx, sb_sml_y2_int, in_mux, net);
+    } else {
+        auto sb_sml_d0 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_SML.P08.D0")));
+        auto sb_sml_y1_int = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_SML.P08.Y1_int")));
+        auto sb_sml_y1 = ctx->getWireByName(IdStringList::concat(x0y0, ctx->idf("SB_SML.P08.Y1")));
+        auto sb_big_y1 = ctx->getWireByName(IdStringList::concat(x2y0, ctx->idf("SB_BIG.P08.Y1")));
+        auto sb_big_ydiag = ctx->getWireByName(IdStringList::concat(x2y0, ctx->idf("SB_BIG.P08.YDIAG")));
+        // x2y2/IM.P08.D0 is x2y0/SB_BIG.P08.Y2
+
+        find_and_bind_downhill_pip(ctx, sb_sml_d0, sb_sml_y1_int, net);
+        find_and_bind_downhill_pip(ctx, sb_sml_y1_int, sb_sml_y1, net);
+        find_and_bind_downhill_pip(ctx, sb_sml_y1, sb_big_y1, net);
+        find_and_bind_downhill_pip(ctx, sb_big_y1, sb_big_ydiag, net);
+        find_and_bind_downhill_pip(ctx, sb_big_ydiag, in_mux, net);
+    }
+
+    route_mult_diag(ctx, net, Loc{loc.x + 1, loc.y + 1, 0}, in_mux, 8);
+}
+} // namespace
+
+NEXTPNR_NAMESPACE_BEGIN
+
+void GateMateImpl::route_mult()
```

Comment:
It would be really nice to have some brief rationale behind this (fixing the routing so the inversion is correct) and ideally sketch ASCII art diagrams or something, so this code isn't quite as opaque to whoever stumbles across it in the future